core/distributed/neighborhood_communicator.cpp

```diff
@@ -0,0 +1,288 @@
+}
+
+
+std::unique_ptr<CollectiveCommunicator>
+NeighborhoodCommunicator::create_inverse() const
+{
+    auto base_comm = this->get_base_communicator();
+    distributed::comm_index_type num_sources;
+    distributed::comm_index_type num_destinations;
+    distributed::comm_index_type weighted;
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Dist_graph_neighbors_count(
+        comm_.get(), &num_sources, &num_destinations, &weighted));
+
+    std::vector<distributed::comm_index_type> sources(num_sources);
+    std::vector<distributed::comm_index_type> destinations(num_destinations);
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Dist_graph_neighbors(
+        comm_.get(), num_sources, sources.data(), MPI_UNWEIGHTED,
+        num_destinations, destinations.data(), MPI_UNWEIGHTED));
+
+    return std::make_unique<NeighborhoodCommunicator>(
+        base_comm, destinations, send_sizes_, send_offsets_, sources,
+        recv_sizes_, recv_offsets_);
+}
+
+
+comm_index_type NeighborhoodCommunicator::get_recv_size() const
+{
+    return recv_offsets_.back();
+}
+
+
+comm_index_type NeighborhoodCommunicator::get_send_size() const
+{
+    return send_offsets_.back();
+}
+
+
+NeighborhoodCommunicator::NeighborhoodCommunicator(
+    communicator base, const std::vector<distributed::comm_index_type>& sources,
+    const std::vector<comm_index_type>& recv_sizes,
+    const std::vector<comm_index_type>& recv_offsets,
+    const std::vector<distributed::comm_index_type>& destinations,
+    const std::vector<comm_index_type>& send_sizes,
+    const std::vector<comm_index_type>& send_offsets)
+    : CollectiveCommunicator(base), comm_(MPI_COMM_NULL)
+{
+    comm_ = create_neighborhood_comm(base, sources, destinations);
+    send_sizes_ = send_sizes;
+    send_offsets_ = send_offsets;
+    recv_sizes_ = recv_sizes;
+    recv_offsets_ = recv_offsets;
+}
+
+
+NeighborhoodCommunicator::NeighborhoodCommunicator(communicator base)
+    : CollectiveCommunicator(std::move(base)),
+      comm_(MPI_COMM_SELF),
+      send_sizes_(),
+      send_offsets_(1),
+      recv_sizes_(),
+      recv_offsets_(1)
+{
+    // ensure that comm_ always has the correct topology
+    std::vector<comm_index_type> non_nullptr(1);
+    non_nullptr.resize(0);
+    comm_ = create_neighborhood_comm(this->get_base_communicator(), non_nullptr,
+                                     non_nullptr);
+}
+
+
+request NeighborhoodCommunicator::i_all_to_all_v(
+    std::shared_ptr<const Executor> exec, const void* send_buffer,
+    MPI_Datatype send_type, void* recv_buffer, MPI_Datatype recv_type) const
+{
+    auto guard = exec->get_scoped_device_id_guard();
+    request req;
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Ineighbor_alltoallv(
+        send_buffer, send_sizes_.data(), send_offsets_.data(), send_type,
+        recv_buffer, recv_sizes_.data(), recv_offsets_.data(), recv_type,
+        comm_.get(), req.get()));
+    return req;
+}
+
+
+std::unique_ptr<CollectiveCommunicator>
```

Comment:
Why return the base type here ? Is it not possible to return an object of type `NeighborhoodCommunicator` ?